norlab_icp_mapper/Mapper.cpp

```diff
@@ -1,61 +1,186 @@
-	radiusFilterParams["dist"] = std::to_string(sensorMaxRange);
+	radiusFilterParams["dist"] = std::to_string(map.getSensorMaxRange());
 	radiusFilterParams["removeInside"] = "0";
 	radiusFilter = PM::get().DataPointsFilterRegistrar.create("DistanceLimitDataPointsFilter", radiusFilterParams);
+
 }
 
-void norlab_icp_mapper::Mapper::loadYamlConfig(const std::string& inputFiltersConfigFilePath, const std::string& icpConfigFilePath,
-											   const std::string& mapPostFiltersConfigFilePath)
+void norlab_icp_mapper::Mapper::validateYamlKeys(const YAML::Node& node, const std::vector<std::string>& validKeys)
 {
-	if(!icpConfigFilePath.empty())
-	{
-		std::ifstream ifs(icpConfigFilePath.c_str());
-		icp.loadFromYaml(ifs);
-		ifs.close();
-	}
-	else
-	{
-		icp.setDefault();
-	}
+    std::unordered_set<std::string> encounteredKeys;
+    if (!node.IsMap())
+    {
+        throw YAML::Exception(node.Mark(), "Expected a YAML Map node.");
+    }
 
-	if(!inputFiltersConfigFilePath.empty())
-	{
-		std::ifstream ifs(inputFiltersConfigFilePath.c_str());
-		inputFilters = PM::DataPointsFilters(ifs);
-		ifs.close();
-	}
+    // Iterate over the map and check if each key is unique
+    for (const auto& kv : node) {
+        const auto key = kv.first.as<std::string>();
 
-	if(!mapPostFiltersConfigFilePath.empty())
-	{
-		std::ifstream ifs(mapPostFiltersConfigFilePath.c_str());
-		mapPostFilters = PM::DataPointsFilters(ifs);
-		ifs.close();
-	}
+        if (encounteredKeys.count(key) > 0)
+        {
+            throw YAML::Exception(kv.second.Mark(), "Duplicated key: " + key);
+        }
+        if (std::find(validKeys.begin(), validKeys.end(), key) == validKeys.end())
+        {
+            throw YAML::Exception(kv.second.Mark(), "Invalid key: " + key);
+        }
+        encounteredKeys.insert(key);
+    }
+}
+
+void norlab_icp_mapper::Mapper::loadYamlConfig(const std::string& configFilePath)
+{
+    std::ifstream ifs(configFilePath.c_str());
+    if (ifs.fail())
+    {
+        throw std::runtime_error("The input config file " + configFilePath + " does not exist");
+    }
+
+    YAML::Node node = YAML::Load(ifs);
+
+    validateYamlKeys(node, std::vector<std::string>{"icp", "input", "post", "mapper"});
+    if (node["icp"])
+    {
+        icp.loadFromYamlNode(node["icp"]);
+    }
+    else
+    {
+        std::cout << "icp config not found, using default" << std::endl;
+        icp.setDefault();
+    }
+
+    if(node["input"])
+    {
+        inputFilters = PM::DataPointsFilters(node["input"]);
+    }
+    else
+    {
+        std::cout << "Input config not found, using empty configuration." << std::endl;
+        // FIXME add default input filters that "just work"
+    }
+
+    if(node["post"])
+    {
+        mapPostFilters = PM::DataPointsFilters(node["post"]);
+    }
+    else
+    {
+        std::cout << "Post config not found, using empty configuration." << std::endl;
+        // FIXME add default post filters that "just work"
```

Comment:
This comment should be removed, having no post filters is fine

Reply:
I disagree; the user can choose not to have any post filters by using an empty input config in his .yaml file.
```yaml
input:
post:
```
Not including any config should be reserved for the default values, similar to libpointmatcher's icp config.